python/ray/tune/ray_trial_executor.py

```diff
@@ -949,20 +949,19 @@ def resource_string(self):
         else:
             return "? CPUs, ? GPUs"
 
-    def on_step_begin(self, trial_runner):
+    def on_step_begin(self, trials: List[Trial]):
         """Before step() called, update the available resources."""
         self._update_avail_resources()
         self._trial_just_finished_before = self._trial_just_finished
         self._trial_just_finished = False
 
-    def on_step_end(self, trial_runner):
+    def on_step_end(self, trials: List[Trial]):
```

Comment:
Those are API breaking changes, though I don't think they are going to impact many users. @richardliaw thoughts?

Reply:
how about let's do a trycatch around these callpoints and tell users that this interface has changed?

Reply:
Hmmm, I really find this distinction between TrialExecutor and RayTrialExecutor very confusing. For one thing, if you look at https://github.com/ray-project/ray/blob/master/python/ray/tune/trial_runner.py#L544, TrialRunner is calling `get_staged_trial()` which is not even in this TrialExecutor interface. 

Unless someone writes their own TrialRunner, or someone  writes their own TrialExecutor and that TrialExecutor somehow happened to implement a method called `get_staged_trial()`, I can't see how this API can be not breaking now but will be breaking after the change.

So my interpretation of all this, this API is really just internal (between our TrialRunner and our RayTrialExecutor). Maybe I am missing something. Happy to take a look together with you guys.

Reply:
We can't really assume that nobody has ever implemented their own TrialExecutor. It's better to be safe than sorry, I feel.

Reply:
It may be broken/people just haven't upgraded/people haven't found it hard to upgrade. Fundamentally, @xwjiang2010 I agree that we should start privatizing this API, but I also agree with @Yard1 that we definitely cannot just make things "disappear" without providing proper signaling (at least 1 release with warning). 

How about let's do the try-catch + raise warnings, open a public RFC, and then adjust the documentation/args to privatize it in 1.8?